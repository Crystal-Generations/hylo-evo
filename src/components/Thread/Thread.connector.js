import { connect } from 'react-redux'
// import { someAction } from 'some/path/to/actions'
import getCurrentUser from 'store/selectors/getCurrentUser'
import thread, { messages } from './sampleData'

// TODO: convert to actions
const fetchAfter = () => {}
const fetchBefore = () => {}
const onThreadPage = () => {}
const offThreadPage = () => {}

export function mapStateToProps (state, props) {
  return {
    currentUser: getCurrentUser(state),
    thread,
    messages,
    pending: false
  }
}

export const mapDispatchToProps = {
  fetchAfter,
  fetchBefore,
  onThreadPage,
  offThreadPage
}

export default connect(mapStateToProps, mapDispatchToProps)
