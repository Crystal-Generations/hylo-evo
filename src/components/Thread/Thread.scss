.thread {
  height: 100vh;
  flex: 1;
  display: flex;
  flex-direction: column;
  position: relative;
}

.header {
  display: flex;
  height: $messages-header-height;
  padding: $messages-header-vertical-padding $messages-header-horizontal-padding;
  border-bottom: $light-border;
  border-left: $light-border;
  border-right: $light-border;
}

.header-text {
  composes: bdy-drk-lg from 'css/typography.scss';
  padding: $space-2x $space-2x $space-2x 0;
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.close-messages {
  cursor: pointer;
  padding: $space-2x 0;

  &:hover {
    text-decoration: none;

    .close-messages-icon {
      color: $color-rhino-60;
    }
  }
}

.more-icon {
  padding: $space-1x 0;
}

.message-form {
  position: absolute;
  bottom: 40px;
  padding: 0 30px;
  width: 100%;
}
