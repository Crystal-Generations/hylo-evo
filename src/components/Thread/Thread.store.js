import { sortBy } from 'lodash/fp'
import { createSelector as ormCreateSelector } from 'redux-orm'
import orm from 'store/models'
import {
  FETCH_THREAD,
  FETCH_BEFORE_MESSAGES,
  FETCH_AFTER_MESSAGES
} from 'store/constants'

export const MODULE_NAME = 'Thread'

// Action Creators
export function fetchThread (threadId) {
  return {
    type: FETCH_THREAD,
    graphql: {
      query: `
        query ($threadId: ID) {
          messageThread (id: $threadId) {
            id
            unreadCount
            lastReadAt
            createdAt
            updatedAt
            participants {
              id
              name
              avatarUrl
            }
            messages(first: 20) {
              id
              createdAt
              text
              creator {
                id
                name
              }
            }
          }
        }
      `,
      variables: {
        threadId
      }
    },
    meta: {
      extractModel: 'MessageThread'
    }
  }
}

export function fetchBeforeMessages (threadId, cursor) {
  return {
    type: FETCH_BEFORE_MESSAGES,
    graphql: {
      query: `
        query ($threadId: ID, $cursor: ID) {
          messageThread (id: $threadId) {
            id
            messages(first: 20, cursor: $cursor) {
              id
              createdAt
              text
              creator {
                id
                name
                avatarUrl
              }
            }
          }
        }
      `,
      variables: {
        threadId,
        cursor
      }
    },
    meta: {
      rootModelName: 'MessageThread'
    }
  }
}

export function fetchAfterMessages () {
  return {
    type: FETCH_AFTER_MESSAGES
  }
}

// Selectors
export const moduleSelector = (state) => state[MODULE_NAME]

export const getThread = ormCreateSelector(
  orm,
  state => state.orm,
  (_, props) => props.threadId,
  (session, threadId) => {
    var thread
    try {
      thread = session.MessageThread.get({id: threadId})
    } catch (e) {
      return null
    }
    return {
      ...thread.ref,
      participants: thread.participants.toModelArray(),
      messages: sortBy(m => new Date(m.createdAt).getTime(),
        thread.messages.toModelArray().map(m => ({...m.ref, creator: m.creator.ref})))
    }
  })
