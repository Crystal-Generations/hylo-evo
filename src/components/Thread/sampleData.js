import faker from 'faker'

export default {
  id: '1',
  lastReadAt: '',

  messagesTotal: 100,
  participants: [
    {id: '1', name: 'Person 1'},
    {id: '2', name: 'Person 2'},
    {id: '7', name: 'Person 7'}
  ]
}

export const messages = [
  {
    id: '1',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '2',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '3',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '4',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '5',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '6',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '7',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '8',
    text: faker.lorem.sentence(),
    creator: {
      id: '1',
      name: 'Person 1',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '9',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  },
  {
    id: '10',
    text: faker.lorem.sentence(),
    creator: {
      id: '2',
      name: 'Person 2',
      avatarUrl: 'https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png'
    },
    createdAt: new Date()
  }
]
