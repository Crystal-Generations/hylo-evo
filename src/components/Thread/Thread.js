import React from 'react'
import { Link } from 'react-router-dom'
import { filter, get, map, min, max, sortBy } from 'lodash/fp'
const { array, bool, func, object, string } = React.PropTypes
import Icon from 'components/Icon'
import MessageSection from 'components/MessageSection'
import MessageForm from 'components/MessageForm'
import PeopleTyping from 'components/PeopleTyping'
import { getSocket, socketUrl } from 'client/websockets'
import './Thread.scss'

export default class Thread extends React.Component {
  static propTypes = {
    id: string,
    currentUser: object,
    thread: object,
    messages: array,
    pending: bool,
    onThreadPage: func,
    offThreadPage: func,
    fetchBefore: func,
    fetchAfter: func
  }

  constructor (props) {
    super(props)
    this.state = {scrolledUp: false}
  }

  setupForThread () {
    return
    const { thread: { id }, onThreadPage } = this.props
    //onThreadPage()
    if (this.socket) {
      this.socket.post(socketUrl(`/noo/post/${id}/subscribe`)) // for people typing

      if (this.reconnectHandler) {
        this.socket.off('reconnect', this.reconnectHandler)
      }

      this.reconnectHandler = () => {
        const { messages, fetchAfter } = this.props
        const afterId = max(map('id', messages))
        fetchAfter(afterId)
        .then(() => this.refs.messageSection.scrollToBottom())
        this.socket.post(socketUrl(`/noo/post/${id}/subscribe`))
      }
      this.socket.on('reconnect', this.reconnectHandler)
    }

    this.refs.form.getWrappedInstance().focus()
  }

  componentDidMount () {
    this.socket = getSocket()
    this.setupForThread()
  }

  disableSocket () {
    const threadId = get('thread.id', this.props)
    if (this.socket) {
      this.socket.off('reconnect', this.reconnectHandler)
      this.socket.post(socketUrl(`/noo/post/${threadId}/unsubscribe`))
    }
  }

  componentWillReceiveProps (nextProps) {
    const oldId = get('thread.id', this.props)
    const newId = get('thread.id', nextProps)
    if (newId !== oldId) this.disableSocket()
  }

  componentDidUpdate (prevProps) {
    const oldId = get('thread.id', prevProps)
    const newId = get('thread.id', this.props)
    if (newId !== oldId && newId) this.setupForThread()
  }

  componentWillUnmount () {
    const { offThreadPage } = this.props
    //offThreadPage()
    this.disableSocket()
  }

  render () {
    const { thread, pending, fetchBefore, currentUser } = this.props
    const { scrolledUp } = this.state
    /*const loadMore = () => {
      if (pending || messages.length >= thread.messagesTotal) return
      const beforeId = min(map('id', messages))
      fetchBefore(beforeId)
      .then(() => this.refs.messageSection.scrollToMessage(beforeId))
    }*/
    const loadMore = () => {}
    const messages = sortBy('createdAt', this.props.messages || [])
    return <div styleName='thread'>
      <Header thread={thread} currentUser={currentUser} />
      <MessageSection {...{messages, pending}} thread={thread}
        onLeftBottom={() => this.setState({scrolledUp: true})}
        onHitBottom={() => this.setState({scrolledUp: false})}
        onScrollToTop={loadMore} ref='messageSection' />
      <div styleName='message-form'>
        <MessageForm threadId={thread.id} ref='form' />
      </div>
      <PeopleTyping showNames showBorder={scrolledUp} />
    </div>
  }
}

function hasNewMessages (messages, thread, currentUser) {
  const latestMessage = messages.length && messages[messages.length - 1]
  const latestFromOther = latestMessage && latestMessage.creator.id !== currentUser.id
  return latestFromOther && thread.lastReadAt &&
    new Date(latestMessage.createdAt) > new Date(thread.lastReadAt)
}

function Header ({ thread, currentUser }) {
  const participants = thread.participants
  const id = get('id', currentUser)
  const others = map('name', filter(f => f.id !== id, participants))
  const othersMinusLast = others.slice(0, others.length - 1)

  return <div styleName='header'>
    <div styleName='header-text'>
      You{others.length > 1 ? `, ${othersMinusLast.join(', ')}` : ''} and {others[others.length - 1]}
    </div>
    <Icon name='More' styleName='more-icon' />
    <Link to='/' styleName='close-messages'>
      <Icon name='Ex' styleName='close-messages-icon' />
    </Link>
  </div>
}
