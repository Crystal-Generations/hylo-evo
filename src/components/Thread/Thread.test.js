import Thread from './Thread'
import { shallow } from 'enzyme'
import React from 'react'

it('matches the last snapshot', () => {
  const match = { params: { threadId: '1' } }
  const wrapper = shallow(<Thread match={match} />)
  expect(wrapper).toMatchSnapshot()
})
