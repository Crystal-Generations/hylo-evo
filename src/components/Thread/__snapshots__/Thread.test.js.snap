// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the last snapshot 1`] = `
<div>
  <Header />
  <Connect(MessageSection)
    messages={Array []}
    onHitBottom={[Function]}
    onLeftBottom={[Function]}
    onScrollToTop={[Function]}
  />
  <div />
  <div>
    <Connect(MessageForm)
      messageThreadId="1"
    />
  </div>
  <PeopleTyping />
</div>
`;
